runtime/dart_isolate.cc

```diff
@@ -280,6 +282,155 @@ std::weak_ptr<DartIsolate> DartIsolate::CreateRootIsolate(
+        return;
+      }
+      old_task_observer_add(key, callback);
+    });
+  };
+
+  UIDartState::Context context(task_runners);
+  context.advisory_script_uri = isolate_group_data->GetAdvisoryScriptURI();
+  context.advisory_script_entrypoint =
+      isolate_group_data->GetAdvisoryScriptEntrypoint();
+  auto isolate_data = std::make_unique<std::shared_ptr<DartIsolate>>(
+      std::shared_ptr<DartIsolate>(
+          new DartIsolate(settings, context, platform_isolate_manager)));
+
+  IsolateMaker isolate_maker =
+      [parent_isolate](
+          std::shared_ptr<DartIsolateGroupData>* unused_isolate_group_data,
+          std::shared_ptr<DartIsolate>* isolate_data, Dart_IsolateFlags* flags,
+          char** error) {
+        return Dart_CreateIsolateInGroup(
+            /*group_member=*/parent_isolate,
+            /*name=*/"PlatformIsolate",
+            /*shutdown_callback=*/
+            reinterpret_cast<Dart_IsolateShutdownCallback>(
+                DartIsolate::SpawnIsolateShutdownCallback),
+            /*cleanup_callback=*/
+            reinterpret_cast<Dart_IsolateCleanupCallback>(
+                DartIsolateCleanupCallback),
+            /*child_isolate_data=*/isolate_data,
+            /*error=*/error);
+      };
+  Dart_Isolate platform_isolate = CreateDartIsolateGroup(
+      nullptr, std::move(isolate_data), nullptr, error, isolate_maker);
+
+  Dart_EnterIsolate(parent_isolate);
+
+  if (*error) {
+    return nullptr;
+  }
+
+  if (!platform_isolate_manager->RegisterPlatformIsolate(platform_isolate)) {
+    // The PlatformIsolateManager was shutdown while we were creating the
+    // isolate. This means that we're shutting down the engine. We need to
+    // shutdown the platform isolate.
+    FML_LOG(INFO) << "Shutdown during platform isolate creation";
+    Dart_ExitIsolate();  // Exit parent_isolate.
+    Dart_EnterIsolate(platform_isolate);
+    Dart_ShutdownIsolate();
+    Dart_EnterIsolate(parent_isolate);
+    return nullptr;
+  }
+
+  Dart_EnterScope();
+  Dart_PersistentHandle entry_point_handle =
+      Dart_NewPersistentHandle(entry_point);
+  Dart_ExitScope();
+
+  platform_task_runner->PostTask([entry_point_handle, platform_isolate,
+                                  weak_platform_isolate_manager]() {
+    std::shared_ptr<PlatformIsolateManager> platform_isolate_manager =
+        weak_platform_isolate_manager.lock();
+    if (platform_isolate_manager == nullptr ||
+        platform_isolate_manager->IsShutdown()) {
+      // Shutdown happened in between this task being posted, and it running.
+      // platform_isolate has already been shut down. Do nothing.
+      FML_LOG(INFO) << "Shutdown before platform isolate entry point";
+      return;
+    }
+    Dart_EnterIsolate(platform_isolate);
+    Dart_EnterScope();
+    Dart_Handle entry_point = Dart_HandleFromPersistent(entry_point_handle);
+    Dart_DeletePersistentHandle(entry_point_handle);
+
+    // Disable Isolate.exit().
```

Comment:
I think this part fits more into setup of the isolate, which is done in the call to `CreateDartIsolateGroup()` above (which calls `InitializeIsolate()` which calls e.g. `LoadLibraries()` which sets up `dart:io` / `dart:ui` / ..)

Consider moving the `Disable Isolate.exit()` above or into `CreateDartIsolateGroup()` (by passing a boolean to it)